Source/Services/ReverseCallDispatcher.cs

```diff
@@ -3,196 +3,300 @@
+        /// <param name="getResponseContex">A delegate to get the <see cref="ReverseCallResponseContext"/> from a <typeparamref name="TResponse"/>.</param>
+        /// <param name="executionContextManager">The <see cref="IExecutionContextManager"/> to use.</param>
+        /// <param name="logger">The <see cref="ILogger"/> to use.</param>
         public ReverseCallDispatcher(
-            IAsyncStreamReader<TResponse> responseStream,
-            IServerStreamWriter<TRequest> requestStream,
-            ServerCallContext context,
-            Expression<Func<TResponse, ulong>> responseProperty,
-            Expression<Func<TRequest, ulong>> requestProperty,
-            ILogger logger)
+                IAsyncStreamReader<TClientMessage> clientStream,
+                IServerStreamWriter<TServerMessage> serverStream,
+                ServerCallContext context,
+                Func<TClientMessage, TConnectArguments> getConnectArguments,
+                Action<TServerMessage, TConnectResponse> setConnectResponse,
+                Action<TServerMessage, TRequest> setMessageRequest,
+                Func<TClientMessage, TResponse> getMessageResponse,
+                Func<TConnectArguments, ReverseCallArgumentsContext> getArgumentsContext,
+                Action<TRequest, ReverseCallRequestContext> setRequestContext,
+                Func<TResponse, ReverseCallResponseContext> getResponseContex,
+                IExecutionContextManager executionContextManager,
+                ILogger logger)
         {
-            _responseStream = responseStream;
-            _requestStream = requestStream;
+            _clientStream = clientStream;
+            _serverStream = serverStream;
             _context = context;
+            _getConnectArguments = getConnectArguments;
+            _setConnectResponse = setConnectResponse;
+            _setMessageRequest = setMessageRequest;
+            _getMessageResponse = getMessageResponse;
+            _getArgumentsContext = getArgumentsContext;
+            _setRequestContext = setRequestContext;
+            _getResponseContex = getResponseContex;
+            _executionContextManager = executionContextManager;
             _logger = logger;
-            _responseProperty = responseProperty.GetPropertyInfo();
-            _requestProperty = requestProperty.GetPropertyInfo();
-
-            _tasks = new[]
-                {
-                    Task.Run(HandleResponse),
-                    Task.Run(HandleResponseProcessing)
-                };
-            _handleResponse = Task.WhenAny(_tasks);
         }
 
         /// <inheritdoc/>
-        public Task Call(TRequest request, Action<TResponse> callback)
+        public TConnectArguments Arguments { get; private set; }
+
+        /// <inheritdoc/>
+        public async Task<bool> ReceiveArguments(CancellationToken cancellationToken)
```

Comment:
Hope this is the last one 😛 But I also think that this should only be called once?

Reply:
Fixed by https://github.com/dolittle-fundamentals/DotNET.Fundamentals/pull/310/commits/8db9b14fce058a43cc0b6c61cbd54de7e3edb160